File: app/src/main/java/com/adityasubathu/vaccinationreminder/LoginActivityLoginFragment.java
package com.adityasubathu.vaccinationreminder;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.PorterDuff;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.text.Editable;
import android.text.TextWatcher;
import android.text.method.PasswordTransformationMethod;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.Objects;

public class LoginActivityLoginFragment extends Fragment implements View.OnClickListener, View.OnTouchListener {

    static String email, password, username, storedUsername, storedPassword;
    EditText usernameField, passwordField;
    TextView signUpLauncherTextView;
    Button loginButton;
    SharedPreferences mySharedPrefs, loginFlagPrefs;
    View v;
    ImageView passwordVisible;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {

        v = inflater.inflate(R.layout.login_fragment, container, false);
        return v;
    }

    @SuppressLint("ClickableViewAccessibility")
    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        usernameField = v.findViewById(R.id.usernameField);
        passwordField = v.findViewById(R.id.passwordField);
        passwordVisible = v.findViewById(R.id.password_login_visibility);

        passwordVisible.setOnTouchListener(this);

        usernameField.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

                usernameField.setCompoundDrawables(null, null, null, null);

                usernameField.getBackground().setColorFilter(getResources().getColor(R.color.colorPrimary), PorterDuff.Mode.SRC_ATOP);
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });
        passwordField.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

                passwordField.setCompoundDrawables(null, null, null, null);

                passwordField.getBackground().setColorFilter(getResources().getColor(R.color.colorPrimary), PorterDuff.Mode.SRC_ATOP);

            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        loginButton = v.findViewById(R.id.loginButton);
        loginButton.setOnClickListener(this);

        signUpLauncherTextView = v.findViewById(R.id.signUpLauncherTextView);
        signUpLauncherTextView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                username = usernameField.getText().toString();
                password = passwordField.getText().toString();

                LoginActivitySignUpFragment signUpFragment = new LoginActivitySignUpFragment();
                FragmentManager fm = Objects.requireNonNull(getActivity()).getSupportFragmentManager();
                FragmentTransaction ft = fm.beginTransaction();
                ft.setCustomAnimations(android.R.anim.fade_in, android.R.anim.fade_out);
                ft.replace(R.id.loginActivityFragmentHolder, signUpFragment, "LoginActivitySignUpFragment");
                ft.commit();
            }
        });

    }

    @Override
    public void onStart() {
        super.onStart();

        if (SplashActivity.publicUsername != null && !SplashActivity.publicUsername.isEmpty()) {

            usernameField.setText(SplashActivity.publicUsername);
            passwordField.requestFocus();
        } else {

            usernameField.requestFocus();
        }

    }

    @Override
    public void onClick(View v) {

        username = usernameField.getText().toString();
        password = passwordField.getText().toString();

        mySharedPrefs = Objects.requireNonNull(getActivity()).getSharedPreferences(username, Context.MODE_PRIVATE);
        loginFlagPrefs = getActivity().getSharedPreferences("loginInfo", Context.MODE_PRIVATE);

        storedUsername = mySharedPrefs.getString("username", null);
        storedPassword = mySharedPrefs.getString("password", null);

        if (username.isEmpty() || password.isEmpty() || !username.equals(storedUsername)) {

            if (username.isEmpty()) {
                changeColorToRed(usernameField, "Please Enter Your Username");
            }

            if (password.isEmpty()) {
                changeColorToRed(passwordField, "Password Cannot Be Empty");
            }

            if (!username.equals(storedUsername)) {
                changeColorToRed(usernameField, "Username Not Registered");
            }

        } else if (username.equals(storedUsername) && !password.equals(storedPassword)) {
            changeColorToRed(passwordField, "Password is Incorrect");
        } else if (username.equals(storedUsername) && password.equals(storedPassword)) {

            SharedPreferences.Editor e = loginFlagPrefs.edit();
            e.putString("activeUser", username);
            LoginActivitySignUpFragment.activeUsername = username;
            e.putBoolean("loggedIn", true);
            e.apply();

            Intent i = new Intent(getActivity(), MainFragmentHolder.class);
            startActivity(i);
            getActivity().finish();
        }

    }

    private void changeColorToRed(@NonNull final EditText field, String errorMessage) {

        //field.setCompoundDrawablesWithIntrinsicBounds(null, null, ContextCompat.getDrawable(Objects.requireNonNull(getActivity()), R.drawable.ic_info_error), null);
        field.getBackground().setColorFilter(getResources().getColor(R.color.error_red), PorterDuff.Mode.SRC_ATOP);
        field.setHintTextColor(getResources().getColor(R.color.error_red));

        field.setError(errorMessage);
        field.requestFocus();

        field.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

                field.setCompoundDrawables(null, null, null, null);

                field.getBackground().setColorFilter(getResources().getColor(R.color.colorPrimary), PorterDuff.Mode.SRC_ATOP);

                field.setHintTextColor(getResources().getColor(R.color.textColorBlack));

            }

            @Override
            public void afterTextChanged(Editable editable) {

            }
        });

    }

    @SuppressLint("ClickableViewAccessibility")
    @Override
    public boolean onTouch(View view, MotionEvent motionEvent) {

        switch (motionEvent.getAction()) {

            case MotionEvent.ACTION_DOWN: {
                passwordField.setTransformationMethod(null);
                passwordField.setSelection(passwordField.getText().length());
                return true;
            }

            case MotionEvent.ACTION_UP: {
                passwordField.setTransformationMethod(PasswordTransformationMethod.getInstance());
                passwordField.setSelection(passwordField.getText().length());
                return false;
            }

        }
        return true;
    }
}

File: app/src/main/java/com/adityasubathu/vaccinationreminder/LoginActivityLoginFragment.kt
package com.adityasubathu.vaccinationreminder

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.text.method.PasswordTransformationMethod
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.View.OnTouchListener
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView
import androidx.fragment.app.Fragment

class LoginActivityLoginFragment : Fragment(), View.OnClickListener, OnTouchListener {
    private lateinit var usernameField: EditText
    private lateinit var passwordField: EditText
    private lateinit var signUpLauncherTextView: TextView
    private lateinit var loginButton: Button
    private lateinit var mySharedPrefs: SharedPreferences
    private lateinit var loginFlagPrefs: SharedPreferences
    private lateinit var v: View
    private lateinit var passwordVisible: ImageView

    var password: String? = null
    lateinit var username: String
    private var storedUsername: String? = null
    private var storedPassword: String? = null


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        v = inflater.inflate(R.layout.login_fragment, container, false)
        return v
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        usernameField = v.findViewById(R.id.usernameField)
        passwordField = v.findViewById(R.id.passwordField)
        passwordVisible = v.findViewById(R.id.password_login_visibility)
        passwordVisible.setOnTouchListener(this)
        usernameField.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                changeColor(usernameField, null, R.color.colorPrimary)
            }

            override fun afterTextChanged(s: Editable) {}
        })
        passwordField.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                changeColor(passwordField, null, R.color.colorPrimary)
            }

            override fun afterTextChanged(s: Editable) {}
        })
        loginButton = v.findViewById(R.id.loginButton)
        loginButton.setOnClickListener(this)
        signUpLauncherTextView = v.findViewById(R.id.signUpLauncherTextView)
        signUpLauncherTextView.setOnClickListener {
            username = usernameField.text.toString()
            password = passwordField.text.toString()
            val signUpFragment = LoginActivitySignUpFragment()
            val fm = activity!!.supportFragmentManager
            val ft = fm.beginTransaction()
            ft.setCustomAnimations(android.R.anim.fade_in, android.R.anim.fade_out)
            ft.replace(R.id.loginActivityFragmentHolder, signUpFragment, "LoginActivitySignUpFragment")
            ft.commit()
        }
    }

    override fun onStart() {
        super.onStart()
        if (SplashActivity.publicUsername.isNotEmpty()) {
            usernameField.setText(SplashActivity.publicUsername)
            passwordField.requestFocus()
        } else {
            usernameField.requestFocus()
        }
    }

    override fun onClick(v: View) {
        username = usernameField.text.toString()
        password = passwordField.text.toString()
        mySharedPrefs = activity!!.getSharedPreferences(username, Context.MODE_PRIVATE)
        loginFlagPrefs = activity!!.getSharedPreferences("loginInfo", Context.MODE_PRIVATE)
        storedUsername = mySharedPrefs.getString("username", null)
        storedPassword = mySharedPrefs.getString("password", null)
        if (username.isEmpty() || password!!.isEmpty() || username != storedUsername) {
            if (username.isEmpty()) {
                changeColor(usernameField, "Please Enter Your Username", R.color.error_red)
            }
            if (password!!.isEmpty()) {
                changeColor(passwordField, "Password Cannot Be Empty", R.color.error_red)
            }
            if (username != storedUsername) {
                changeColor(usernameField, "Username Not Registered", R.color.error_red)
            }
        } else if (username == storedUsername && password != storedPassword) {
            changeColor(passwordField, "Password is Incorrect", R.color.error_red)
        } else if (username == storedUsername && password == storedPassword) {
            val e = loginFlagPrefs.edit()
            e.putString("activeUser", username)
            LoginActivitySignUpFragment.activeUsername = username
            e.putBoolean("loggedIn", true)
            e.apply()
            val i = Intent(activity, MainFragmentHolder::class.java)
            startActivity(i)
            activity!!.finish()
        }
    }

    private fun changeColor(field: EditText, errorMessage: String?, color: Int) {
        field.background.setTint(color)
        field.setHintTextColor(color)
        if (errorMessage!=null) {
            field.error = errorMessage
        }
        field.requestFocus()
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouch(view: View, motionEvent: MotionEvent): Boolean {
        when (motionEvent.action) {
            MotionEvent.ACTION_DOWN -> {
                passwordField.transformationMethod = null
                passwordField.setSelection(passwordField.text.length)
                return true
            }
            MotionEvent.ACTION_UP -> {
                passwordField.transformationMethod = PasswordTransformationMethod.getInstance()
                passwordField.setSelection(passwordField.text.length)
                return false
            }
        }
        return true
    }
}